### PR TITLE
Change default logic to ALL for all SMT2 solvers and allow cvc4 to support Array of Bool (Issue5977 quick fix)

### PR DESCRIPTION
**Pull request by TGWDB**
This is a quick fix for [Issue 5977](https://github.com/diffblue/cbmc/issues/5977) that has some probably issues elsewhere.

The main changes are:

1. Set the default logic to `ALL` for all SMT2 solvers.
2. Allow `cvc4` translation to support `Array` of `Bool` (instead of `(_ BitVec 1)`.

The potential problems with the first modification are that using `ALL` logics may potentially be slower, and also that for `z3` (and other solvers?) this can result in an `unknown` outcome instead of `sat` or `unsat`. At the moment `cbmc` does **NOT** support `unknown` results and this causes an error.

There is not an obvious problem with the second change, except that this has not been significantly tested.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by TGWDB** [comment]
This PR has taken a blunt approach and set the logic for to `ALL` for every SMT2 solver. The logic can be set to `ALL` for only `cvc4` by adding an additional line
```
    logic = "ALL";
```
around line 99 of `src/solvers/smt2/smt2_conv.cpp`. This would limit the scope of the change in a potentially more safe manner, but also creates more inconsistency between different solvers. For now this does not appear an important choice if we just want to get things working with `cvc4` for the upcoming deadline, but for merging this fix into `develop` a good decision and justification is necessary.

**Comment by SaswatPadhi** [comment]
CVC4 used to raise a warning on `(set-logic ALL)`. Instead it required `(set-logic ALL_SUPPORTED)`: https://cvc4.github.io/datatypes.html#logic

CVC5 now accepts `(set-logic ALL)` though: https://github.com/cvc5/cvc5/blob/master/NEWS#L53

> Removed support for redundant logics ALL_SUPPORTED and QF_ALL_SUPPORTED,
  use ALL and QF_ALL instead.

**Comment by TGWDB** [comment]
> CVC4 used to raise a warning on `(set-logic ALL)`. Instead it required `(set-logic ALL_SUPPORTED)`: https://cvc4.github.io/datatypes.html#logic
> 
> CVC5 now accepts `(set-logic ALL)` though: https://github.com/cvc5/cvc5/blob/master/NEWS#L53
> 
> > Removed support for redundant logics ALL_SUPPORTED and QF_ALL_SUPPORTED,
> > use ALL and QF_ALL instead.

Thanks for the detail. I am not seeing any reports of warnings on using `ALL` locally and so have not pushed a change, but this can be done (for `cvc4` only in `smt2_conv.cpp` if necessary).

I've just pushed a change that treats `unknown` as `unsat`. This may simplify some of the ideas you were suggesting by email. Some output below (note that I have not posted the actual example since I am not sure if it can be shared, but it is the one sent by email), as usual some detail is cut (including properties to not share code information):

```
tgw@DB0907:~/github/testing/AWS-202107-deadline$ /home/tgw/github/playground-cbmc-build/bin/cbmc --cvc4 example01.c
CBMC version 5.32.1 (cbmc-5.32.1-105-gf9be0ebfb-dirty) 64-bit x86_64 linux
...
Passing problem to SMT2 ALL using CVC4
...
** 0 of 3 failed (1 iterations)
VERIFICATION SUCCESSFUL
```
So this now handles the arrays of bitvectors as bools and also `unknown` results are treated as `unsat` yielding the same outcome as `z3`.

Note that this is WIP and may NOT be sound or a good idea!

File: src/goto-checker/solver_factory.cpp [hunk 1]
```diff
@@ -357,7 +357,7 @@ solver_factoryt::get_smt2(smt2_dect::solvert solver)
       ns,
       "cbmc",
       std::string("Generated by CBMC ") + CBMC_VERSION,
-      "QF_AUFBV",
+      "ALL",
```
Comment (thomasspriggs) [inline src/goto-checker/solver_factory.cpp:360]: :pick: The title of this commit states that the change is only being make for `cvc4`, but the code changes appear to be changing it for all solvers.

File: src/solvers/smt2/smt2_conv.cpp [hunk 2]
```diff
@@ -96,6 +96,7 @@ smt2_convt::smt2_convt(
     break;
 
   case solvert::CVC4:
+    use_array_of_bool = true;
```
Comment (thomasspriggs) [inline src/solvers/smt2/smt2_conv.cpp:99]: :pick: The change of logic used and the switch on of `use_array_of_bool` appear to be two separate changes which should be in two separate commits. Is `use_array_of_bool` enough to fix the problem on its own? I think the error returned from the solver was to do with invalid sorts rather than unavailable theories.

File: src/solvers/smt2/smt2_dec.cpp [hunk 3]
```diff
@@ -148,32 +148,32 @@ decision_proceduret::resultt smt2_dect::read_result(std::istream &in)
 
     const auto &parsed = parsed_opt.value();
 
-    if(parsed.id()=="sat")
-      res=resultt::D_SATISFIABLE;
-    else if(parsed.id()=="unsat")
-      res=resultt::D_UNSATISFIABLE;
+    if(parsed.id() == "sat")
+      res = resultt::D_SATISFIABLE;
+    else if(parsed.id() == "unsat" || parsed.id() == "unknown")
```
Comment (thomasspriggs) [inline src/solvers/smt2/smt2_dec.cpp:153]: :pick: Seems like a fairly dubious change best left in the `smt-fast-response` branch only, rather than added to develop.

File: src/solvers/smt2/smt2_dec.cpp [hunk 4]
```diff
@@ -148,32 +148,32 @@ decision_proceduret::resultt smt2_dect::read_result(std::istream &in)
...
+      res = resultt::D_UNSATISFIABLE;
     else if(
       parsed.id().empty() && parsed.get_sub().size() == 1 &&
       parsed.get_sub().front().get_sub().size() == 2)
     {
-      const irept &s0=parsed.get_sub().front().get_sub()[0];
-      const irept &s1=parsed.get_sub().front().get_sub()[1];
+      const irept &s0 = parsed.get_sub().front().get_sub()[0];
+      const irept &s1 = parsed.get_sub().front().get_sub()[1];
```
Comment (thomasspriggs) [inline src/solvers/smt2/smt2_dec.cpp:160]: :pick: This formatting commit noise, would need to be cleaned up before PRing into `develop`. I am not going to block this PR based on this issue, for a non-develop quick fix PR.

**Review by thomasspriggs — APPROVED** [review]
